src/primitives/block.h

```diff
@@ -123,27 +123,56 @@ class CBlock : public CBlockHeader
         return block;
     }
 
-    /** Compute at which vout of the block's coinbase transaction the witness commitment occurs, or -1 if not found */
-    int GetWitnessCommitmentIndex() const
+    /**
+     * Get the vout index of the segwit commitment in the coinbase transaction of this block.
+     *
+     * Returns -1 if no witness commitment was found.
+     */
+    inline int GetWitnessCommitmentIndex() const
     {
-        int commitpos = NO_WITNESS_COMMITMENT;
-        if (!vtx.empty()) {
-            for (size_t o = 0; o < vtx[0]->vout.size(); o++) {
-                const CTxOut& vout = vtx[0]->vout[o];
-                if (vout.scriptPubKey.size() >= MINIMUM_WITNESS_COMMITMENT &&
-                    vout.scriptPubKey[0] == OP_RETURN &&
-                    vout.scriptPubKey[1] == 0x24 &&
-                    vout.scriptPubKey[2] == 0xaa &&
-                    vout.scriptPubKey[3] == 0x21 &&
-                    vout.scriptPubKey[4] == 0xa9 &&
-                    vout.scriptPubKey[5] == 0xed) {
-                    commitpos = o;
-                }
+        return vtx.empty() ? NO_WITNESS_COMMITMENT : GetWitnessCommitmentIndex(*vtx.at(0));
+    }
+
+    /**
+     * Get the vout index of the segwit commitment in the given coinbase transaction.
+     *
+     * Returns -1 if no witness commitment was found.
+     */
+    template<typename T> static inline int GetWitnessCommitmentIndex(const T& coinbase) {
```

Comment:
I still don't understand why this is moved https://github.com/bitcoin/bitcoin/pull/16411#discussion_r335017902

* GetWitnessCommitmentIndex is a member function, but doesn't access any member variables. This is confusing and makes it hard to write clean code.
* `src/primitives` only contains data structures, and no validation logic. If this becomes the new place for validation, we might as well move the rest of validation to this file. E.g. ConnectBlock to `CBlock::Connect`, etc. This clearly defeats the purpose of having different modules for primitives and validation.

Reply:
style-nit: member functions are already inline, same goes for `template<>`s. No need to specify it twice.

Reply:
I hear you on the "only data structure" part... perhaps this belongs in validation.cpp in the end. It's bloated and I didn't wanna bloat it more with the new witness commitment section stuff, though.

Reply:
On the other hand, it could be argued that this is in fact data structure *stuff* and not validation. I mean, there's no validation done here, it simply locates and the index of the witness commitment, if any.

Reply:
`GetWitnessCommitmentIndex` has been added to validation in commit 8b49040854be2e26b66366aeae1cba4716f93d93. Generally we don't move stuff around unless there is a reason to. Maybe @sipa can give more insights. Though, if no one else complains about this, I will close this conversation as resolved.

However, making this a member function of CBlock, whereas it doesn't read any of the member variables, is highly confusing and should be fixed.

Reply:
I strongly prefer keeping this in validation. Not every piece of code that uses transactions (e.g. wallets, P2P layer things, ...) needs to depend on something that only matters to validation.

Reply:
I've restored it to validation.cpp, and added a variant that takes a transaction rather than a block. The block-based one (original) is converted into a wrapper calling the variant.